docs: Add security policy. (#1262)
slsa-framework/slsa-github-generator: SECURITY.md

> # Security Policy

Author: Maybe we should eventually roll this up into some kind of overall slsa-framework or OpenSSF-wide security policy/reporting process?

@MarkLodato @laurentsimon @asraa @joshuagl Thoughts?

Author: At the very least a lot of this should be shared between slsa-github-generator and slsa-verifier.

Reviewer: org-level makes sense to me too! We could file an issue on https://github.com/slsa-framework/governance to track

Author: Added issue https://github.com/slsa-framework/governance/issues/19 to track.

[Report a vulnerability](https://github.com/slsa-framework/slsa-github-generator/security/advisories/new)

Please see
[Privately reporting a security vulnerability](https://docs.github.com/en/code-security/security-advisories/guidance-on-reporting-and-writing/privately-reporting-a-security-vulnerability#privately-reporting-a-security-vulnerability)
for more information on how to submit a vulnerability using GitHub's interface.

### When Should I Report a Vulnerability?

- You think you discovered a potential security vulnerability in slsa-github-generator
- You are unsure how a vulnerability affects slsa-github-generator
- You think you discovered a vulnerability in another project that slsa-github-generator depends on
  - For projects with their own vulnerability reporting and disclosure process, please report it directly there

### When Should I NOT Report a Vulnerability?

- You need help tuning GitHub Actions for security
- You need help applying security related updates
- Your issue is not security related

### Vulnerability Response

Each report is acknowledged and analyzed by the [Security Team](#security-team)

> within 14 days. This will set off the

Author: 14 days seems like a long time to just acknowledge a report but that was the limit for the "initial response time" in the OpenSSF best practices (assuming "initial response time" meant acknowledgement of the vulnerability report and not public disclosure) and I didn't want to necessarily promise more than that.

@laurentsimon @asraa @joshuagl Thoughts?

Reviewer 2: I'm fine with 14 days to start with.

Reviewer: Agreed, makes sense to start with a larger value and reduce over time as we gain confidence.

[Security Release Process](#security-release--disclosure-process).

Any vulnerability information shared with the Security Team stays within
slsa-github-generator project and will not be disseminated to other projects
unless it is necessary to get the issue fixed.

As the security issue moves from triage, to identified fix, to release planning
we will keep the reporter updated.

## Security Release & Disclosure Process

Security vulnerabilities should be handled quickly and sometimes privately. The
primary goal of this process is to reduce the total time users are vulnerable
to publicly known exploits.

### Private Disclosure

We ask that all suspected vulnerabilities be privately and responsibly
disclosed via the [private disclosure process](#reporting-a-vulnerability)
outlined above.

Fixes may be developed and tested by the [Security Team](#security-team) in

> private repositories if deemed necessary.

Reviewer: Do we mean private from the reporter? Or the GitHub Security Advisory private branches which are private from the general public? I think the latter should be the default way we handle fixes for a reported vulnerability.

Author: Yes, I meant the latter. I'll clarify.

Author: Done.

### Public Disclosure

> Vulnerabilities are disclosed publicly as GitHub

Reviewer: ```suggestion
Resolved vulnerabilities are disclosed publicly as GitHub
```
?

Author: I think there may be instances where we disclose vulnerabilities if there is an existing mitigation. But also in the worst case if a vulnerability is made public before the fact we may need to disclose it without one. So there are instances where we disclose without it necessarily being resolved.

and the bug submitter. We prefer to fully disclose the bug as soon as possible
once a user mitigation is available. It is reasonable to delay disclosure when
the bug or the fix is not yet fully understood, the solution is not
well-tested, or for vendor coordination. The timeframe for disclosure is from
immediate (especially if it's already publicly known) to several weeks. For a
vulnerability with a straightforward mitigation, we expect report date to
disclosure date to be on the order of 14 days.

If you know of a publicly disclosed security vulnerability please IMMEDIATELY
[report the vulnerability](#reporting-a-vulnerability) to inform the
[Security Team](#security-team) about the vulnerability so they may start the
patch, release, and communication process.

If possible the Security Team will ask the person making the public report if
the issue can be handled via a private disclosure process. If the reporter
denies the request, the Security Team will move swiftly with the fix and
release process. In extreme cases you can ask GitHub to delete the issue but
this generally isn't necessary and is unlikely to make a public disclosure less
damaging.

> ### Security Releases

Reviewer 3: Slack is probably the best place, or maybe eventually via e-mail if we have enough major users

> ### Security Releases

Reviewer: > We should probably get in the habit of announcing new releases in more places anyway. regardless of security issues...

💯

> ### Security Releases

Author: Added a note about announcing  to the slack channel.

Once a fix is available it will be released and announced via the project on
GitHub. Security releases will announced and clearly marked as a security
release and include information on which vulnerabilities were fixed. As much as
possible this announcement should be actionable, and include any mitigating
steps users can take prior to upgrading to a fixed version.

Fixes will be applied in patch releases to all [supported
versions](#supported-versions).

### Severity

The [Security Team](#security-team) evaluates vulnerability severity on a
case-by-case basis, guided by [CVSS 3.1](https://www.first.org/cvss/v3.1/specification-document).

> ## Security Posture

Author: There is currently nothing here about how we manage repo owners/admins and settings. I'm not sure it needs to be in the initial version but given that folks are referencing and executing code directly from the repo it's worth writing down at some point.

Maybe as part of writing a formal contributor ladder/onboarding process that we mostly spitballed for Naveen.
https://github.com/slsa-framework/slsa-github-generator/pull/428#issuecomment-1194931713

Reviewer: Agree that it would be very useful to capture this, let's file an issue to ensure we don't forget?

Author: https://github.com/slsa-framework/slsa-github-generator/issues/1291

We aim to reduce the number of security issues through several general
security-concious development practices including the use of unit-tests,
end-to-end (e2e) tests, static and dynamic analysis tools, and use of
memory-safe languages.

We aim to fix issues discovered by analysis tools as quickly as possible. We
prefer to add these tools to "pre-submit" checks on PRs so that issues are
never added to the code in the first place.

> In general, we follow the following security-concious practices during

Reviewer: ```suggestion
In general, we observe the following security-conscious practices during
```

development (This is not an exhaustive list).

- All PRs are reviewed by at least one [CODEOWNER](./CODEOWNERS).
- All unit and linter pre-submit tests must pass before a PRs is merged. See
  the [Pre-submits and Unit Tests](./CONTRIBUTING.md#pre-submits-and-unit-tests)
  section of the Contributor Guide for more information.
- All releases include no known e2e test failures. See
  [RELEASE.md](./RELEASE.md) for info on the release process. See the
  [End to End (e2e) Tests](./CONTRIBUTING.md#end-to-end-e2e-tests) section of
  the Contributor Guide for more information on e2e tests.
- We refrain from using memory-unsafe languages (e.g. C, C++) or memory-unsafe
  use of languages that are memory-safe by default (e.g. the Go
  [unsafe](https://pkg.go.dev/unsafe) package).

## Security Team

The Security Team is responsible for the overall security of the
project and for reviewing reported vulnerabilities. Each member is familiar
with designing secure software, security issues related to CI/CD, GitHub

> actions and build provenance.

Reviewer: ```suggestion
Actions and build provenance.
```

You can view current team membership here:

> [slsa-framework/slsa-github-generator-security](https://github.com/orgs/slsa-framework/teams/slsa-github-generator-security/members)

Author: I created this team and I intend to initially add the 4 current repo admins. However, I wanted input on this PR before spamming invites to the GitHub team.

Reviewer 2: I'm fine with the team.

Reviewer: AIUI teams and their members are only visible to members of the organisation, so that URL won't be useful to most folks reading this document.

https://docs.github.com/en/organizations/organizing-members-into-teams/changing-team-visibility

Author: hrm, well that means that it won't be useful for this purpose. I guess we can just add names here.

Author: I added a list of names since a team can't be viewed externally to the org.

> Security Team membership is currently considered on a case-by-case basis.

Author: We should also lay out the process for joining the SRC/VMT/Security team at some point. Similar to the contributor ladder.

Reviewer: ➕

Author: I'll leave this hanging for now. We'll revisit once we have an org-wide policy.
